packages/react-reconciler/src/ReactFiberLane.js

```diff
@@ -7,16 +7,677 @@
+
+  const expiredLanes = root.expiredLanes;
+  const suspendedLanes = root.suspendedLanes;
+  const pingedLanes = root.pingedLanes;
+
+  // Check if any work has expired.
+  if (expiredLanes !== NoLanes) {
+    nextLanes = expiredLanes;
+    nextLanePriority = return_highestLanePriority = SyncLanePriority;
+    equalOrHigherPriorityLanes = (getLowestPriorityLane(nextLanes) << 1) - 1;
+  } else {
+    // Do not work on any idle work until all the non-idle work has finished,
+    // even if the work is suspended.
+    const nonIdlePendingLanes = pendingLanes & NonIdleLanes;
+    if (nonIdlePendingLanes !== NoLanes) {
+      const nonIdleUnblockedLanes = nonIdlePendingLanes & ~suspendedLanes;
+      if (nonIdleUnblockedLanes !== NoLanes) {
+        nextLanes = getHighestPriorityLanes(nonIdleUnblockedLanes);
+        nextLanePriority = return_highestLanePriority;
+        equalOrHigherPriorityLanes = (1 << return_updateRangeEnd) - 1;
+      } else {
+        const nonIdlePingedLanes = nonIdlePendingLanes & pingedLanes;
+        if (nonIdlePingedLanes !== NoLanes) {
+          nextLanes = getHighestPriorityLanes(nonIdlePingedLanes);
+          nextLanePriority = return_highestLanePriority;
+          equalOrHigherPriorityLanes = (1 << return_updateRangeEnd) - 1;
+        }
+      }
+    } else {
+      // The only remaining work is Idle.
+      const unblockedLanes = pendingLanes & ~suspendedLanes;
+      if (unblockedLanes !== NoLanes) {
+        nextLanes = getHighestPriorityLanes(unblockedLanes);
+        nextLanePriority = return_highestLanePriority;
+        equalOrHigherPriorityLanes = (1 << return_updateRangeEnd) - 1;
+      } else {
+        if (pingedLanes !== NoLanes) {
+          nextLanes = getHighestPriorityLanes(pingedLanes);
+          nextLanePriority = return_highestLanePriority;
+          equalOrHigherPriorityLanes = (1 << return_updateRangeEnd) - 1;
+        }
+      }
+    }
+  }
+
+  if (nextLanes === NoLanes) {
+    // This should only be reachable if we're suspended
+    // TODO: Consider warning in this path if a fallback timer is not scheduled.
+    return NoLanes;
+  }
+
+  // If there are higher priority lanes, we'll include them even if they
+  // are suspended.
+  nextLanes = pendingLanes & equalOrHigherPriorityLanes;
+
+  // If we're already in the middle of a render, switching lanes will interrupt
+  // it and we'll lose our progress. We should only do this if the new lanes are
+  // higher priority.
+  const wipLanes = root.wipLanes;
+  if (
+    wipLanes !== NoLanes &&
+    wipLanes !== nextLanes &&
+    // If we already suspended with a delay, then interrupting is fine. Don't
+    // bother waiting until the root is complete.
+    (wipLanes & suspendedLanes) === NoLanes
+  ) {
+    getHighestPriorityLanes(wipLanes);
+    const wipLanePriority = return_highestLanePriority;
+    if (nextLanePriority <= wipLanePriority) {
+      return wipLanes;
+    } else {
+      return_highestLanePriority = nextLanePriority;
```

Comment:
Yeah probably should for consistency

Reply:
I guess since it's only used by the inner function, it's fine to leave it.

Might get rid of the "register" thing. Too clever. If only we knew someone on TC39 who could advocate for stack allocated tuples!